core/internal/monitor/cwmetadata.go

```diff
@@ -56,20 +55,20 @@
 	// GraphQL client to communicate with the W&B backend.
 	graphqlClient graphql.Client
 
-	// W&B entity to use with the gql.OrganizationCoreWeaveOrganizationID query.
-	entity string
-
 	// Internal debug logger.
 	logger *observability.CoreLogger
 
-	// The scheme and hostname for contacting the metadata server,
-	// not including a final slash. For example, "http://localhost:8080".
-	baseURL *url.URL
-
-	// The relative path on the server to which to make requests.
+	// settings contain the info needed to probe the CoreWeave metadata.
 	//
-	// This must not include the schema and hostname prefix.
-	endpoint string
+	// Specifically:
+	//  - Entity is W&B entity to use with the gql.OrganizationCoreWeaveOrganizationID query.
+	//    May be updated upon a run start with the information from the server, which is
+	//    the main reason we keep a pointer to a Settings struct.
```

Comment:
Note: I do want Settings to eventually become immutable, so that they're completely determined by `wandb.init()`. This is fine for now though as I don't have an alternative for getting the resolved entity.

There's also a subtle data race risk: `Settings` doesn't use locks, so this implicitly relies on calls to the update and getter methods being synchronized in other ways. It works because we don't call Probe() until after Sender responds to the RunStartRequest, but this is a very hard-to-spot dependency.

Reply:
I added a mutex to the settings struct -- using it for every settings that has a setter (rather, a corresponding UpdateX method).